torch/csrc/jit/passes/onnx/function_extraction.cpp

```diff
@@ -0,0 +1,1098 @@
+    void DebugPrint() const;
+    void SetAttrName(Node* ref_n, Symbol attr, const std::string& name);
+    void SetAttrName(Node* ref_const_n, const std::string& name);
+    c10::optional<std::string> FindAttrName(Node* ref_n, Symbol attr);
+    c10::optional<std::string> FindAttrName(Node* ref_const_n);
+
+    ScopePtr scope_key_;
+    scope_ctx_map scope_ctxs_;
+    std::unordered_map<Node*, std::unordered_set<Node*>> constant_map_;
+    std::unordered_map<
+        Node*,
+        std::unordered_map<Symbol, std::unordered_set<Node*>>>
+        attribute_map_;
+
+    // Passed later to serialization.
+    ValAttrNameMap val_attr_to_name_;
+    NodeAttrNameMap node_attr_to_name_;
+  };
+
+  using FunctionCtxPtr = FunctionContext*;
+  using func_ctx_map = std::unordered_map<ScopePtr, FunctionCtxPtr>;
+
+  static bool IsValidScope(ScopePtr s);
+  static c10::optional<ScopePtr> InferScope(Node* n);
+  static bool IsAncestor(ScopePtr parent, ScopePtr child);
+  static c10::optional<ScopePtr> FindCommonAncestor(ScopePtr a, ScopePtr b);
+  static c10::optional<ScopePtr> FindCommonAncestor(const scope_list& scopes);
+  std::shared_ptr<Graph> ConstructFuncGraph(FunctionContext& ctx);
+
+  void ConvertScopeToFunction(
+      const ScopePtr& scope_key,
+      const scope_list& scope_list,
+      scope_ctx_map& scope_ctxs,
+      const std::shared_ptr<Graph>& graph);
+
+  static void HandleNoScopeNodes(scope_ctx_map&, node_list no_scope_nlist);
+  std::tuple<scope_ctx_map, node_list> PartitionNodesByScope(Block* b);
+  scope_ctx_map PartitionNodesByScope(const std::shared_ptr<Graph>& graph);
+  static std::unordered_map<ScopePtr, scope_list> PartitionIdenticalScopes(
+      scope_ctx_map& scope_ctxs);
+  static scope_list SortScopesByMaxDepth(
+      std::unordered_map<ScopePtr, scope_list>&);
+  Node* CreateFunctionDefNode(
+      FunctionContext& func_ctx,
+      const std::shared_ptr<Graph>& graph,
+      const std::string& domain_name,
+      const std::string& func_name);
+  Node* CreateFunctionNode(
+      FunctionContext& func_ctx,
+      ScopeContext& scope_ctx,
+      const std::shared_ptr<Graph>& graph,
+      const std::string& domain_name,
+      const std::string& func_name);
+
+  static void DebugPrintScopeContexts(const scope_ctx_map&);
+  static void DebugPrintGraphWithFunction(const std::shared_ptr<Graph>& g);
+  static void DebugPrintConstantDiff(const FunctionContext&);
+
+  std::shared_ptr<Graph> graph_;
+  std::unordered_set<std::string> module_names_;
+  std::unordered_set<std::string> param_names_;
+  // Track modules with same module name that are exported as different onnx
+  // local functions.
+  std::unordered_map<std::string, int> module_variant_count_;
+  func_ctx_map func_ctxs_;
+};
+
+FunctionExtractor::FunctionContext::FunctionContext(
+    ScopePtr key,
+    const scope_list& scopes,
+    scope_ctx_map& scope_ctxs)
+    : scope_key_(key) {
+  GRAPH_UPDATE(
+      "Process function family for scope ",
```

Comment:
where is the term "function family" defined?
If it's not defined anywhere, I suggest defining it in the top-of-file comment I asked for.